packages/ui/src/models/camel/pipe-resource.ts

```diff
@@ -16,25 +21,46 @@ export class PipeResource implements CamelResource {
       return;
     }
     this.pipe = json as PipeType;
+    this.flow = new PipeVisualEntity(this.pipe.spec?.source, this.pipe.spec?.steps, this.pipe.spec?.sink);
+    this.metadata = new MetadataEntity(this.pipe.metadata);
+    this.errorHandler = new PipeErrorHandlerEntity(this.pipe.spec?.errorHandler);
   }
 
   getEntities(): BaseCamelEntity[] {
-    return []; // TODO [new MetadataEntity(this.pipe.metadata), new ErrorHandlerEntity(this.pipe.spec?.errorHandler)];
+    const answer = [];
+    if (this.metadata?.metadata) {
+      answer.push(this.metadata);
+    }
+    if (this.errorHandler?.errorHandler) {
+      answer.push(this.errorHandler);
+    }
+    return answer;
   }
 
   getType(): SourceSchemaType {
     return SourceSchemaType.Pipe;
   }
 
   getVisualEntities(): PipeVisualEntity[] {
-    return [new PipeVisualEntity(this.pipe.spec?.source, this.pipe.spec?.steps, this.pipe.spec?.sink)];
+    return this.flow ? [this.flow] : [];
   }
 
   supportsMultipleVisualEntities(): boolean {
     return false;
   }
 
   toJSON(): PipeType {
+    this.pipe = {
+      ...this.pipe,
+      metadata: this.metadata?.toJSON().metadata,
+      spec: {
+        ...this.pipe.spec,
+        errorHandler: this.errorHandler?.toJSON().errorHandler,
+        source: this.flow.source,
+        steps: this.flow.steps,
+        sink: this.flow.sink,
+      },
+    };
     return this.pipe;
```

Comment:
[question] Is there any other way to do this? Mutating an object as a side effect of serializing seems counterintuitive, ideally, this should be an idempotent operation.

Reply:
Like passing callback function from `CamelResource` to the entities? we can do that. I think this way is not so uncommon tho, it's kinda lazy loading, gathering the update when it's really needed.

Reply:
Fixed as we discussed, let entities update the object through the reference.